backtracking/graph_coloring.cpp

```diff
@@ -1,72 +1,107 @@
-#include <stdio.h>
+/**
+ * @file
+ * @brief prints the assigned colors
+ */
+#include <iostream>
+#include <array>
+#include <vector>
 
-// Number of vertices in the graph
-#define V 4
-
-void printSolution(int color[]);
-
-/* A utility function to check if the current color assignment
-   is safe for vertex v */
-bool isSafe(int v, bool graph[V][V], int color[], int c) {
-    for (int i = 0; i < V; i++)
-        if (graph[v][i] && c == color[i])
-            return false;
-    return true;
-}
+/**
+ * @namespace
+ * @brief Backtracking algorithms
+ */
+namespace backtracking {
+    /** A utility function to print solution
+     * @param color description
+     * @param V number of vertices in the graph
+     */
+    void printSolution(const std::array <int, 4>& color, int V) {
+        std::cout << "Following are the assigned colors\n";
+        for (int i = 0; i < V; i++) {
+            std::cout << color[i];
+        }
+        std::cout << "\n";
+    }
 
-/* A recursive utility function to solve m coloring problem */
-void graphColoring(bool graph[V][V], int m, int color[], int v) {
-    /* base case: If all vertices are assigned a color then
-       return true */
-    if (v == V) {
-        printSolution(color);
-        return;
+    /** A utility function to check if the current color assignment is safe for
+     * vertex v
+     * @tparam V number of vertices in the graph
+     * @param v description
+     * @param graph description
+     * @param color description
+     * @param c description
+     * @returns `true` if ....
+     * @returns `false` if ....
+     */
+    template <size_t V>
+    bool isSafe(int v, const std::array<std::array <int, V>, V>& graph, const std::array <int, 4> &color, int c) {
+        for (int i = 0; i < V; i++) {
+            if (graph[v][i] && c == color[i]) {
+                return false;
+            }
+        }
+        return true;
     }
 
-    /* Consider this vertex v and try different colors */
-    for (int c = 1; c <= m; c++) {
-        /* Check if assignment of color c to v is fine*/
-        if (isSafe(v, graph, color, c)) {
-            color[v] = c;
+    /** A recursive utility function to solve m coloring problem
+     * @tparam V number of vertices in the graph
+     * @param graph description
+     * @param m description
+     * @param [in,out] color description // used in,out to notify in documentation
+     * that this parameter gets modified by the function
+     * @param v description
```

Comment:
Same here as [above](https://github.com/TheAlgorithms/C-Plus-Plus/pull/916#discussion_r458253263).

Reply:
> Same here as [above](https://github.com/TheAlgorithms/C-Plus-Plus/pull/916#discussion_r458253263).

The pull-request should be marked as draft or work in progress until all information is available to the author. Otherwise, it is a waste of time and effort for the reviewers to review half-complete pull-requests with known issues.